### PR TITLE
Adds sub-command clear-session to history command. Issue #5791

### PR DESCRIPTION
**Pull request by s7rul**
## Description
Adds sub-command ``clear-session`` to ``history`` command. Which clear current session history. If ``history merge`` or ``builtin history merge`` is run in a session only the history after this will be erased.

## Implementation
A function ``clear_session`` witch does this is added to ``src/history.cpp`` where it first puts all of the commands in ``new_items`` into ``deleted_items``after this ``new_items`` is cleared and ``first_unwritten_new_item_index`` is reset.

Documentation is updated in ``doc_src`` and completion. Tests were written for the added code (no unit test in ``fish_tests.cpp``).

Fixes issue #5791

## Additional notes
We have done this as part in a school project and it is our first contribution to open-source.

## TODO

- [x] Make test pass
- [x] Make it not affect old history via de-duplication.

### CODE REVIEW

**Comment by ridiculousfish** [comment]
This is great! Impressively simple implementation, nice tests too. I am able to reproduce the test failures locally, I will investigate to ensure we aren't masking some serious bug.

Regarding touching old history, faho is correct that this will erase items from previous sessions, if the item is also in the current session. I think this can be fixed easily by tracking in `deleted_items` whether the deletion is for all sessions or just this session, and then inspect that field when de-duplicating to decide whether to erase. If you'd like to tackle that please do; otherwise I will have a go and add on top of this commit.

**Comment by s7rul** [comment]
> Regarding touching old history, faho is correct that this will erase items from previous sessions, if the item is also in the current session. I think this can be fixed easily by tracking in `deleted_items` whether the deletion is for all sessions or just this session, and then inspect that field when de-duplicating to decide whether to erase. If you'd like to tackle that please do; otherwise I will have a go and add on top of this commit.

I can give it a try, sounds simple enough.

**Comment by ridiculousfish** [comment]
So the test failures were due to [this hack](https://github.com/fish-shell/fish-shell/blob/8784253282289455ae2168be6eb27232cfdbb1d9/src/history.cpp#L575) and not a problem with this PR.

The gory details: fish shares a single history file with all other instances.  Any instance should only "see" past items, not items from other running instances. It does this via timestamps: a fish instance has a start timestamp. Items with earlier timestamps are "old" and should be considered part of history. Items with later timestamps are "new" and should be ignored by this instance.

A problem arises with items whose timestamp exactly matches fish's start date: those items may be old or new. For example `exec fish` is likely to create an "old" history item that exactly matches fish's start date. Meanwhile scripted tests also create history items in the same second as fish launches. To distinguish these, any new history item created in the same second as fish starts will have its timestamp incremented by one.

Really we should be using higher resolution timestamps. I'll see if I can fix this another way but it shouldn't block merge.

**Comment by ridiculousfish** [comment]
> Also perhaps `history clear --session` instead?

I prefer a separate command `clear-session`. Usually you want commands to do the least dangerous thing and for modifiers to make them more dangerous, e.g. `rm` vs `rm -R`. But with `history clear --session` it would be reversed: the default would be more destructive. So I think a separate command is better in this case.

File: src/history.cpp [hunk 1]
```diff
@@ -1073,6 +1076,15 @@ void history_impl_t::clear() {
     this->clear_file_state();
 }
 
+void history_impl_t::clear_session() {
+    for (size_t i = 0; i < new_items.size(); i++) {
+        deleted_items.insert(new_items.at(i).str());
```
Comment (ridiculousfish) [inline src/history.cpp:1081]: nit: you could use a range-based for loop here to simplify it:

    for (const auto &item : new_items) {
        deleted_items.insert(item.str());
    }

File: share/completions/history.fish [hunk 2]
```diff
@@ -1,5 +1,5 @@
 # Note that when a completion file is sourced a new block scope is created so `set -l` works.
-set -l __fish_history_all_commands search delete save merge clear
+set -l __fish_history_all_commands search delete save merge clear clear-session
```
Comment (zx8) [inline share/completions/history.fish:2]:
```suggestion
set -l __fish_history_all_commands search delete save merge clear clear_session
```

File: share/functions/history.fish [hunk 3]
```diff
@@ -62,13 +62,15 @@ function history --description "display or manipulate interactive command histor
         set hist_cmd search
     else if set -q _flag_merge
         set hist_cmd merge
+    else if set -q _flag_clear-session
```
Comment (zx8) [inline share/functions/history.fish:65]:
```suggestion
    else if set -q _flag_clear_session
```

File: src/history.cpp (hunk 1 above)
Comment (s7rul) [reply src/history.cpp]: Changed this in later commit.

**Comment by ridiculousfish** [comment]
Looks good, squash-merged as 049104e8df02a989e3c0ad70d2521251192013cd. Thank you, a wonderful contribution!

**Pull request closed without merging**